src/Orleans/CodeGeneration/GenericMethodInvoker.cs

```diff
@@ -0,0 +1,164 @@
+
+            // Next, create a delegate which will call the method on the grain, pushing each of the arguments,
+            var il = new ILDelegateBuilder<GenericMethodInvokerDelegate>(
+                FieldBuilder,
+                $"GenericMethodInvoker_{this.grainInterfaceType}_{concreteMethod.Name}",
+                GenericMethodInvokerDelegateMethodInfo);
+
+            // Load the grain and cast it into the appropriate type.
+            il.LoadArgument(0);
+            il.CastOrUnbox(this.grainInterfaceType);
+
+            // Load every argument from the argument array.
+            var methodParameters = concreteMethod.GetParameters();
+            for (var i = 0; i < methodParameters.Length; i++)
+            {
+                il.LoadArgument(1); // Load the argument array.
+
+                // The argument is offset by the type parameter count, since type parameters come first.
+                il.LoadConstant(i + this.typeParameterCount); 
+                il.LoadReferenceElement();
+                il.CastOrUnbox(methodParameters[i].ParameterType);
+            }
+
+            // Call the target method.
+            il.Call(concreteMethod);
+
+            // If the result type is Task or Task<T>, convert it to Task<object>.
+            var returnType = concreteMethod.ReturnType;
+            if (returnType != typeof(Task<object>))
+            {
+                var boxMethod = BoxMethods.GetOrAdd(returnType, CreateBoxMethod);
+                il.Call(boxMethod);
+            }
+
+            il.Return();
+            return il.CreateDelegate();
+        }
+
+        private static MethodInfo GetBoxMethod(Type returnType)
+        {
+            if (returnType == typeof(Task)) return TypeUtils.Method((Task task) => task.Box());
+            if (returnType == typeof(void)) return TypeUtils.Property(() => TaskDone.Done).GetMethod;
+
+            if (returnType.GetGenericTypeDefinition() != typeof(Task<>))
+                throw new ArgumentException($"Unsupported return type {returnType}.");
+            var innerType = returnType.GenericTypeArguments[0];
+            var methods = typeof(PublicOrleansTaskExtensions).GetMethods(BindingFlags.Static | BindingFlags.Public);
+            foreach (var method in methods)
+            {
+                if (method.Name != "Box" || !method.ContainsGenericParameters) continue;
+                return method.MakeGenericMethod(innerType);
+            }
+
+            throw new ArgumentException($"Could not find Box method for type {returnType}");
+        }
+
+        bool IEqualityComparer<object[]>.Equals(object[] x, object[] y)
+        {
+            if (ReferenceEquals(x, y)) return true;
+            if (ReferenceEquals(x, null)) return false;
+            if (ReferenceEquals(null, y)) return false;
+
+            if (x.Length < this.typeParameterCount || y.Length < this.typeParameterCount) return false;
```

Comment:
In practice we expect x & y to have the same lengths, but this equality comparison is only checking the prefix of those arrays, because the intention is to check the type parameters only & not the entire contents. The array is specified as  type params followed by method arguments